Add integration for Oracle Cloud Infrastructure (OCI) Generative AI

raveharpaz opened the pull request:
# Description
Oracle Cloud Infrastructure (OCI) Generative AI is a fully managed service that provides a set of state-of-the-art large language models (LLMs) which are available through a single API. This PR adds two new Integration packages (llms and embeddings) and relevant documentation for using OCI Generative AI services with LlamaIndex.

Fixes # NA

## New Package? Yes
llama-index-llms-oci-genai
llama_index.embeddings.oci_genai

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

nerdai commented on llama-index-integrations/embeddings/llama-index-embeddings-oci-genai/pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+import_path = "llama_index.embeddings.oci_genai"
+
+[tool.llamahub.class_authors]
+OCIGenAIEmbeddings = "llama-index"
```
feel free to use your github username instead of `llama-index` here

raveharpaz replied:
addressed

nerdai commented on llama-index-integrations/llms/llama-index-llms-oci-genai/pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+import_path = "llama_index.llms.oci_genai"
+
+[tool.llamahub.class_authors]
+OCIGenAI = "llama-index"
```
feel free to use your github username instead of `llama-index` here

raveharpaz replied:
addressed

nerdai commented on llama-index-integrations/llms/llama-index-llms-oci-genai/pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+
+[tool.poetry]
+authors = ["Arthur Cheng <arthur.cheng@oracle.com>", "Rave Harpaz <rave.harpaz@oracle.com>"]
+description = "llama-index llms OCI GenAI integration"
```
please add `exclude = ["**/BUILD"]` below `description`

raveharpaz replied:
addressed

nerdai commented on llama-index-integrations/embeddings/llama-index-embeddings-oci-genai/pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+
+[tool.poetry]
+authors = ["Arthur Cheng <arthur.cheng@oracle.com>", "Rave Harpaz <rave.harpaz@oracle.com>"]
+description = "llama-index embeddings oci-genai integration"
```
please add `exclude = ["**/BUILD"]` below `description`

raveharpaz replied:
addressed

nerdai commented on llama-index-integrations/embeddings/llama-index-embeddings-oci-genai/README.md:
Please include a detailed README for your package. Our integrations are discoverable via https://llamahub.ai and the README is displayed there are and is useful for prospective users.

raveharpaz replied:
addressed, hopefully the right level of detail is now provided

nerdai commented on llama-index-integrations/llms/llama-index-llms-oci-genai/README.md:
Please include a detailed README for your package. Our integrations are discoverable via https://llamahub.ai and the README is displayed there are and is useful for prospective users.

raveharpaz replied:
addressed

nerdai commented on llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:
```diff
@@ -0,0 +1,347 @@
+
+        response = self._client.chat(request)
+
+        return ChatResponse(
```
ooc: does this support LogProbs? If so, can we include them?

nerdai reviewed (commented):
Thanks for the contribution @raveharpaz! I left some minor comments in my review.

Looks like our lint checks our failing. Please run `make lint` and `make format`, commit changes and then push.

raveharpaz commented:
@nerdai thank you very much for the review. 
All of your comments have been addressed with exception the last (feature not currently supported). 
We look forward to the results of the next round :-) 
Thanks again,
Rave

nerdai commented:
Thanks @raveharpaz looks like lint check still not passing.

Can you please run make lint and make format, commit changes and then push.

raveharpaz commented on llama-index-integrations/embeddings/llama-index-embeddings-oci-genai/llama_index/embeddings/oci_genai/base.py:
```diff
@@ -0,0 +1,266 @@
+        return self._embed(text, input_type="SEARCH_DOCUMENT")
+
+    async def _aget_text_embedding(self, text: str) -> List[float]:
+        return self._get_text_embedding(text)
```
correct, client does NOT currently have proper async support, so we are doing synchronous instead

raveharpaz commented:
Dear @nerdai , @logan-markewich , we have the addressed the points you brought up. 
Please let us know if there is anything else, otherwise thanks again very much and looking forward to seeing our integration merged :-)
Rave
